### PR TITLE
Write connection file in jupyter dir if present && shorten log output

### PR DESCRIPTION
**Pull request by tinmarino**
Fix part2 of #2:
The connection file are written in the current directory which could be annoying.
If jupyter is present, it can give a default data path.
ipy_thread_filename

You can see that the ls (top right) do not show any `kernel-*.json` and the `--existing` advice (bottom left) is shortened to the only required pid (proof right).
Also, you can see print as no return because it is another PR to follow the Minimum Viable Change philosophy

### CODE REVIEW

File: kernel.py [hunk 1]
```diff
@@ -152,9 +159,15 @@ def _write_connection_file(self):
         # The key should be secret, to only allow the same user to connect.
         # Make sure the permissions are set accordingly.
         os.chmod(self.connection_filename, os.stat(self.connection_filename).st_mode & 0o0700)
+
+        def shorten_filename(runtime_file):
+            """Shorten connection filename kernel-24536.json -> 24536"""
+            r_cfile = r'.*kernel-([^\-]*).*\.json'
+            return re.sub(r_cfile, r'\1', runtime_file)
+
```
Comment (albertz) [inline kernel.py]: This `shorten_filename` is only correct if it used the `jupyter_runtime_dir`, right?
Also, what if `connection_filename` had another pattern?

File: kernel.py [hunk 2]
```diff
@@ -79,9 +80,15 @@ def __init__(self, connection_filename="kernel.json", connection_fn_with_pid=Tru
         self._lock = threading.Lock()
         self._condition = threading.Condition(lock=self._lock)
 
+        # Get connection file path
         if connection_fn_with_pid:
             name, ext = os.path.splitext(connection_filename)
             connection_filename = "%s-%i%s" % (name, os.getpid(), ext)
+            try:
+                from jupyter_core.paths import jupyter_runtime_dir
+                connection_filename = os.path.join(jupyter_runtime_dir(), connection_filename)
```
Comment (albertz) [inline kernel.py]: I think we should not add another directory on `connection_filename` if the user provided some own custom `connection_filename`.
Rather, we should change the default `connection_filename` to `None`, and then in that case, if also `connection_fn_with_pid`, then we can have this logic, which falls back to the old default `connection_filename = "kernel.json"`.

**Review by albertz — CHANGES_REQUESTED** [review]
Thanks for that. However, I would better change it a bit. See my comments.

File: kernel.py (hunk 1 above)
Comment (tinmarino) [reply kernel.py]: If there is no match no replacement.
The danger is some `../some_dir/anystuff1-kernel-anystuff2.json`. So to correct that, we must keep an other member variable to know if we are on default  conf (i.e. None)

File: kernel.py [hunk 3]
```diff
@@ -105,6 +97,22 @@ def __init__(self, connection_filename="kernel.json", connection_fn_with_pid=Tru
             logger.addHandler(logging.StreamHandler(sys.stdout))
         self._logger = logger
 
+    def _get_connection_filename(self, connection_filename, connection_fn_with_pid):
+        """Craft connection file path"""
```
Comment (albertz) [inline kernel.py]: There should be a newline after and before `"""`. And the parameters and the return type should be documented. I.e. like:
```
"""
:param str connection_filename:
:param bool connection_fn_with_pid:
:return: crafted connection file path
:rtype: str
"""
```

File: kernel.py [hunk 4]
```diff
@@ -105,6 +97,22 @@ def __init__(self, connection_filename="kernel.json", connection_fn_with_pid=Tru
...
+        if connection_fn_with_pid and connection_filename is None:
+            self._should_reduce_filename = True
```
Comment (albertz) [inline kernel.py]: Such a function should not have side effects, like setting this attrib. You would not expect that from the function name.
But maybe easier, just do not create such a separate function at all.

File: kernel.py [hunk 5]
```diff
@@ -223,7 +234,8 @@ def _start_kernel(self):
         self._setup_streams()
         self._create_kernel()
 
-        self._logger.info("IPython: Start kernel now. pid: %i, thread: %r" % (os.getpid(), threading.current_thread()))
+        self._logger.info("IPython: Start kernel now. pid: %i, thread: %r",
+                          os.getpid(), threading.current_thread())
```
Comment (albertz) [inline kernel.py]: Why this change?
Also, better do not indent like this. The indent should not be so long.

File: kernel.py (hunk 1 above)
Comment (albertz) [reply kernel.py]: Yes, that's what I mean.
Btw, also, isn't it easier to just not introduce `shorten_filename` here, and instead have the code directly? It's much shorter and thus easier to understand.

File: kernel.py (hunk 4 above)
Comment (albertz) [reply kernel.py]: Also, this is wrong here. This should come after the `os.path.join` part. Only for that case, we should set `_should_reduce_filename = True`.

Comment (tinmarino) [reply kernel.py]: > Also, this is wrong here. This should come after the `os.path.join` part. Only for that case, we should set `_should_reduce_filename = True`.

Well seen: the user may not have jupyter (and its path and its connection file searching routine). It was a (well hidden) error.
Fix: 4914afb70bc

File: kernel.py (hunk 3 above)
Comment (tinmarino) [reply kernel.py]: Fix: 4914afb70bc (see below)

File: kernel.py (hunk 5 above)
Comment (tinmarino) [reply kernel.py]: This change (`+` -> `,`) was advice by my linter to get some lazy string concatenation.
I agree with the pending indent.
And as sparse is better than dense, I propose that: 99f148334

File: kernel.py (hunk 1 above)
Comment (tinmarino) [reply kernel.py]: Yes inline is more readable. Finally the function got pretty short.
Prettify : c824bcf89

File: kernel.py [hunk 6]
```diff
@@ -79,11 +80,8 @@ def __init__(self, connection_filename="kernel.json", connection_fn_with_pid=Tru
         self._lock = threading.Lock()
         self._condition = threading.Condition(lock=self._lock)
 
-        if connection_fn_with_pid:
-            name, ext = os.path.splitext(connection_filename)
-            connection_filename = "%s-%i%s" % (name, os.getpid(), ext)
-        self.connection_filename = connection_filename
-
+        self.connection_filename, self._should_reduce_filename = self._craft_connection_filename(
+            connection_filename, connection_fn_with_pid)
```
Comment (albertz) [inline kernel.py:105]: Maybe better directly also inline this? I think it becomes more readable then.

File: kernel.py (hunk 5 above)
Comment (albertz) [reply kernel.py]: Yes, that's ok.

File: kernel.py (hunk 6 above)
Comment (tinmarino) [reply kernel.py:105]: I did not do so because I prefer not to charge the `__init__` method as every reader must start there. Maybe they don't care about connection file creation (15 lines).

Comment (albertz) [reply kernel.py:105]: Ok.

**Comment by albertz** [comment]
Sorry, there is a conflict now after I merged the other PR in. Can you rebase? Otherwise I think we can also merge this in now.

**Comment by tinmarino** [comment]
Sure, give me 5 min.
This is the problem of Minimum viable change method (of witch I am not really fan). Are you actually ?

**Review by albertz — APPROVED** [review]

**Comment by tinmarino** [comment]
Tested and pushed.
The only real conflict was the `__init__` parameters. But git got trouble because some function were added just close to some changes.

**Comment by albertz** [comment]
Thanks, merged now.
I guess sometimes Git (or diff/patching) could be more clever, like in such a case, although it can be tricky, and sometimes not so straightforward how it actually should resolve.

**Comment by tinmarino** [comment]
A pleasure (not to say an honor), to have collaborated. Have a nice Sunday.

**Comment by albertz** [comment]
I have to thank you. :)
Btw, I also just pushed a new release to [Pypi](https://pypi.org/project/background_zmq_ipython/).
